File: contests/WC253/Q2.py
from typing import List
from heapq import *


class Solution:
    def minStoneSum(self, piles: List[int], k: int) -> int:
        piles = list(-p for p in piles)
        heapify(piles)

        for _ in range(k):
            x = heappop(piles)
            heappush(piles, x//2)

        return -sum(piles)

File: contests/WC253/Q2.ts
import { MaxPriorityQueue } from '@datastructures-js/priority-queue'

function minStoneSum(piles: number[], k: number): number {
  const que = new MaxPriorityQueue<number>({ priority: (x) => x })
  piles.forEach((p) => que.enqueue(p))
  ;[...Array(k)].forEach((_) => {
    const { element } = que.dequeue()
    que.enqueue(Math.ceil(element / 2))
  })

  let sum = 0
  while (!que.isEmpty()) {
    const { element } = que.dequeue()
    sum += element
  }
  return sum
}
